Add support for toolchain java 14

davido opened the pull request:
Closes #11017.

Test Plan:

1. Create java_tools from this commit:

```bash
  $ bazel build src:java_tools_java14.zip
```

2. Switch to using the java_tools from the above step in WORKSPACE file:

```python
  load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
  http_archive(
    name = "remote_java_tools_linux",
    urls = [
        "file:///<path to the java_tools_java14.zip>",
    ],
  )
```

3. Add Zulu OpenJDK14 to use as javabase in WORKSPACE file:

```python
  http_archive(
    name = "openjdk14_linux_archive",
    build_file_content = """
java_runtime(name = 'runtime', srcs =  glob(['**']), visibility = ['//visibility:public'])
exports_files(["WORKSPACE"], visibility = ["//visibility:public"])
""",
    sha256 = "48bb8947034cd079ad1ef83335e7634db4b12a26743a0dc314b6b861480777aa",
    strip_prefix = "zulu14.28.21-ca-jdk14.0.1-linux_x64",
    urls = ["https://cdn.azul.com/zulu/bin/zulu14.28.21-ca-jdk14.0.1-linux_x64.tar.gz"],
  )
```

4. Activate custom java toolchain and javabase in .bazelrc file:

```bash
  build --java_toolchain=@remote_java_tools_linux//:toolchain_jdk_14
  build --host_java_toolchain=@remote_java_tools_linux//:toolchain_jdk_14
  build --javabase=@openjdk14_linux_archive//:runtime
  build --host_javabase=@openjdk14_linux_archive//:runtime
```

5. Create Java 14 example class file:

```java
  public class Javac14Example {

    record Point(int x, int y) {}

    public static void main(String[] args) {
      Point point = new Point(0, 1);
      System.out.println(point.x);
    }
  }
```

6. Add Bazel file to build Java 14 syntax class with activated preview
   features:

```python
  java_binary(
    name = "Javac14Example",
    srcs = ["Javac14Example.java"],
    javacopts = ["--enable-preview"],
    jvm_flags = ["--enable-preview"],
    main_class = "Javac14Example",
  )
```

7. Test that it works as expected:

```
  $ bazel run java:Javac14Example
INFO: Analyzed target //java:Javac14Example (1 packages loaded, 2 targets configured).
INFO: Found 1 target...
INFO: From Building java/Javac14Example.jar (1 source file):
Note: java/Javac14Example.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
Target //java:Javac14Example up-to-date:
  bazel-bin/java/Javac14Example.jar
  bazel-bin/java/Javac14Example
INFO: Elapsed time: 1.502s, Critical Path: 1.30s
INFO: 1 process: 1 worker.
INFO: Build completed successfully, 2 total actions
INFO: Build completed successfully, 2 total actions
0
```

jin commented:
cc @lberki

lberki commented:
Thanks for the month of patience! We should really have reviewed this earlier but... things came up.

comius commented:
Dear @davido, sorry for the waiting. I did a quick general review of your pull request and at first sight it seems this is a work in progress. 

Did you plan to do additional work on it? Do you need some support with it?

Also may I ask what was the reasoning to drop Java 12 support?

Thanks again for your effort.

davido commented:
> Dear @davido, sorry for the waiting. I did a quick general review of your pull request and at first sight it seems this is a work in progress.

Right. It depends on my Error Prone 2.4.0 PR. Because currently used EP doesn't support Java 14 yet.
 
> Did you plan to do additional work on it? Do you need some support with it?

Yes, I will resume the work on this PR, once EP upgrade is landed.

> Also may I ask what was the reasoning to drop Java 12 support?

In fact we could preserve Java 12. but what is the point? Java 12 and 13 are dead.

> Thanks again for your effort.

One thing so solve is to conditionally remove these two lines in toolchain definition:

```diff
 # override the javac in the JDK.
        #"--patch-module=java.compiler=$(location :java_compiler_jar)",
        #"--patch-module=jdk.compiler=$(location :jdk_compiler_jar)",
```

Do you have a suggestion how to do it: preserve those lines for Java 8, 11 and 12 toolchains, but rmove them for 14, 15, 17, ...?

One idea I had, is to duplicate this toolchain, and name it differently, say: `toolchain_new_jdk_14`  or something? Because right now this PR broke toolchains for Java 8, 11 and 12 (unconditionally removing the above lines).

comius commented:
> In fact we could preserve Java 12, but what is the point? Java 12 and 13 are dead.

Well one thing I can think of is, there might be people out there with projects depending on Java 12. Having the option and not being forced into switch might help them.

> 
> One thing so solve is to **conditionally** remove these two lines in toolchain definition for java toolchans >= 14:
> 
> ```diff
>  # override the javac in the JDK.
>         #"--patch-module=java.compiler=$(location :java_compiler_jar)",
>         #"--patch-module=jdk.compiler=$(location :jdk_compiler_jar)",
> ```
> 
> Do you have a suggestion how to do it: preserve those lines for Java 8, 11 and 12 toolchains, but remove them for 14, 15, 17, ...?
> 
> One idea I had, is to duplicate this toolchain, and name it differently, say: `toolchain_new_jdk_(14|15|17)` or something? Because right now this PR broke toolchains for Java 8, 11 and 12 (unconditionally removing the above lines).

You can use inline ```if``` there, something like ``` ] + ( [] if JavaVersion >= 12 else ["--patch1", "--patch2"]) + [``` or even better declare a variable before the rule: 
```
if JavaVersion >= 12: 
  _PATCH=[] 
else: 
  _PATCH = ["p1","p2"]

java_toolchain(
  ...
  jvm_args = [...] + _PATCH + [...]
```

How to actually properly write the condition for java version or if it is even possible I don't know. If you cannot write it then yes, your best option is to split into two toolchains.

davido commented:
@comius Thanks @philwo, EP 2.4.0 upgrade was merged today and new `java_tools` release was conducted.

I will resume the work here, Also note, that Java 14 toolchain is removed now: https://github.com/bazelbuild/bazel/pull/11666.

I should probably wait until the above PR is merged, because this PR touched the same code area.

davido commented:
Some tests are failing, e.g.:

https://buildkite.com/bazel/bazel-bazel-github-presubmit/builds/6456#98c62ab3-76df-4257-9ec3-0d15cc9803c4

because download doesn't work:

https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/98c62ab3-76df-4257-9ec3-0d15cc9803c4/src/test/shell/bazel/bazel_java_test_jdk14_prebuilt_toolchain_head/attempt_1.log

```
Analyzing: target //java/main:main (1 packages loaded, 0 targets configured)
INFO: Repository openjdk14_linux_archive instantiated at:
  no stack (--record_rule_instantiation_callstack not enabled)
Repository rule http_archive defined at:
  /var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/sandbox/linux-sandbox/5285/execroot/io_bazel/_tmp/d618688373d8c1dc5ac3a5dc5a7bcac8/root/c8d20c307803200fe58c026c171542fc/external/bazel_tools/tools/build_defs/repo/http.bzl:336:31: in <toplevel>
WARNING: Download from https://cdn.azul.com/zulu/bin/zulu14.28.21-ca-jdk14.0.1-linux_x64.tar.gz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException Unknown host: cdn.azul.com
ERROR: An error occurred during the fetch of repository 'openjdk14_linux_archive':
   java.io.IOException: Error downloading [https://cdn.azul.com/zulu/bin/zulu14.28.21-ca-jdk14.0.1-linux_x64.tar.gz] to /var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/sandbox/linux-sandbox/5285/execroot/io_bazel/_tmp/d618688373d8c1dc5ac3a5dc5a7bcac8/root/c8d20c307803200fe58c026c171542fc/external/openjdk14_linux_archive/zulu14.28.21-ca-jdk14.0.1-linux_x64.tar.gz: Unknown host: cdn.azul.com
ERROR: /var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/sandbox/linux-sandbox/5285/execroot/io_bazel/_tmp/d618688373d8c1dc5ac3a5dc5a7bcac8/root/c8d20c307803200fe58c026c171542fc/external/bazel_tools/tools/jdk/BUILD:66:26: @bazel_tools//tools/jdk:legacy_current_java_runtime depends on @openjdk14_linux_archive//:runtime in repository @openjdk14_linux_archive which failed to fetch. no such package '@openjdk14_linux_archive//': java.io.IOException: Error downloading [https://cdn.azul.com/zulu/bin/zulu14.28.21-ca-jdk14.0.1-linux_x64.tar.gz] to /var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/sandbox/linux-sandbox/5285/execroot/io_bazel/_tmp/d618688373d8c1dc5ac3a5dc5a7bcac8/root/c8d20c307803200fe58c026c171542fc/external/openjdk14_linux_archive/zulu14.28.21-ca-jdk14.0.1-linux_x64.tar.gz: Unknown host: cdn.azul.com
ERROR: Analysis of target '//java/main:main' failed; build aborted: Analysis failed
INFO: Elapsed time: 0.147s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (17 packages loaded, 57 targets configured)
FAILED: Build did NOT complete successfully (17 packages loaded, 57 targets configured)
```

Cc @aiuto @philwo @meteorcloudy 

Can you help to upload the required artifact to bazel mirror?

This test is working as expected here locally:

```
  $ bazel test src/test/shell/bazel/bazel_java_test_jdk14_prebuilt_toolchain_head
DEBUG: /home/davido/.cache/bazel/_bazel_davido/4596c3304267e4c44bca87d41e038f29/external/bazel_toolchains/rules/rbe_repo/version_check.bzl:59:14: 
Current running Bazel is not a release version and one was not defined explicitly in rbe_autoconfig target. Falling back to '3.1.0'
DEBUG: /home/davido/.cache/bazel/_bazel_davido/4596c3304267e4c44bca87d41e038f29/external/bazel_toolchains/rules/rbe_repo/checked_in.bzl:103:14: rbe_ubuntu1804_java11 not using checked in configs as detect_java_home was set to True 
DEBUG: /home/davido/.cache/bazel/_bazel_davido/4596c3304267e4c44bca87d41e038f29/external/bazel_toolchains/rules/rbe_repo/version_check.bzl:59:14: 
Current running Bazel is not a release version and one was not defined explicitly in rbe_autoconfig target. Falling back to '3.1.0'
DEBUG: /home/davido/.cache/bazel/_bazel_davido/4596c3304267e4c44bca87d41e038f29/external/bazel_toolchains/rules/rbe_repo/checked_in.bzl:103:14: rbe_ubuntu1604_java8 not using checked in configs as detect_java_home was set to True 
DEBUG: /home/davido/.cache/bazel/_bazel_davido/4596c3304267e4c44bca87d41e038f29/external/build_bazel_rules_nodejs/internal/common/check_bazel_version.bzl:47:14: 
Current Bazel is not a release version, cannot check for compatibility.
DEBUG: /home/davido/.cache/bazel/_bazel_davido/4596c3304267e4c44bca87d41e038f29/external/build_bazel_rules_nodejs/internal/common/check_bazel_version.bzl:49:14: Make sure that you are running at least Bazel 0.21.0.
DEBUG: /home/davido/.cache/bazel/_bazel_davido/4596c3304267e4c44bca87d41e038f29/external/build_bazel_rules_nodejs/internal/common/check_bazel_version.bzl:47:14: 
Current Bazel is not a release version, cannot check for compatibility.
DEBUG: /home/davido/.cache/bazel/_bazel_davido/4596c3304267e4c44bca87d41e038f29/external/build_bazel_rules_nodejs/internal/common/check_bazel_version.bzl:49:14: Make sure that you are running at least Bazel 0.21.0.
INFO: Analyzed target //src/test/shell/bazel:bazel_java_test_jdk14_prebuilt_toolchain_head (9 packages loaded, 224 targets configured).
INFO: Found 1 test target...
Target //src/test/shell/bazel:bazel_java_test_jdk14_prebuilt_toolchain_head up-to-date:
  bazel-bin/src/test/shell/bazel/bazel_java_test_jdk14_prebuilt_toolchain_head
INFO: Elapsed time: 141.694s, Critical Path: 141.28s
INFO: 1 process: 1 linux-sandbox.
INFO: Build completed successfully, 5 total actions
//src/test/shell/bazel:bazel_java_test_jdk14_prebuilt_toolchain_head     PASSED in 140.9s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 5 total actions
```

davido commented:
When trying to build Gerrit with java_tools produced by this PR the log is flooded with:

```
INFO: From Compiling Java headers javatests/com/google/gerrit/server/query/group/libabstract_query_tests-hjar.jar (1 source file):
OpenJDK 64-Bit Server VM warning: Options -Xverify:none and -noverify were deprecated in JDK 13 and will likely be removed in a future release.
INFO: From Compiling Java headers javatests/com/google/gerrit/server/query/project/libabstract_query_tests-hjar.jar (1 source file):
OpenJDK 64-Bit Server VM warning: Options -Xverify:none and -noverify were deprecated in JDK 13 and will likely be removed in a future release.
INFO: From Compiling Java headers javatests/com/google/gerrit/server/query/change/libabstract_query_tests-hjar.jar (2 source files):
OpenJDK 64-Bit Server VM warning: Options -Xverify:none and -noverify were deprecated in JDK 13 and will likely be removed in a future release.
```

This is because of: https://bugs.openjdk.java.net/browse/JDK-8214731.

Should `-Xverify:none` be removed from Bazel entirely?

meteorcloudy commented:
https://mirror.bazel.build/cdn.azul.com/zulu/bin/zulu14.28.21-ca-jdk14.0.1-linux_x64.tar.gz should be available now.

davido commented:
> 
> 
> https://mirror.bazel.build/cdn.azul.com/zulu/bin/zulu14.28.21-ca-jdk14.0.1-linux_x64.tar.gz should be available now.

Thanks for the quick upload. Could you also upload the tars for Darwin and Wndows? Thanks!

meteorcloudy commented:
https://mirror.bazel.build/cdn.azul.com/zulu/bin/zulu14.28.21-ca-jdk14.0.1-macosx_x64.tar.gz
https://mirror.bazel.build/cdn.azul.com/zulu/bin/zulu14.28.21-ca-jdk14.0.1-win_x64.zip
Sure!

davido commented:
Still cannot fetch external resource failure:

```
ERROR: /var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/sandbox/linux-sandbox/4894/execroot/io_bazel/_tmp/6d8619f665a5626e6d9dbba4da652f3a/root/29acb6a64a3aa283b71d49b24e4365c0/external/bazel_tools/tools/jdk/BUILD:66:26: @bazel_tools//tools/jdk:legacy_current_java_runtime depends on @openjdk14_linux_archive//:runtime in repository @openjdk14_linux_archive which failed to fetch. no such package '@openjdk14_linux_archive//': java.io.IOException: Error downloading [https://mirror.bazel.build/cdn.azul.com/zulu/bin/zulu14.28.21-ca-jdk14.0.1-linux_x64.tar.gz] to /var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/sandbox/linux-sandbox/4894/execroot/io_bazel/_tmp/6d8619f665a5626e6d9dbba4da652f3a/root/29acb6a64a3aa283b71d49b24e4365c0/external/openjdk14_linux_archive/zulu14.28.21-ca-jdk14.0.1-linux_x64.tar.gz: Unknown host: mirror.bazel.build
ERROR: Analysis of target '//:test' failed; build aborted: Analysis failed
```

I guess the CI is running in a sandbox and is not allowed to fetch anything from the outside, and i have to provide the new JDK 14 runtime as separate resources for each OS?

davido commented:
Thanks @philwo, we have this FAQ entry: [1], and even this example fix: [2].
Apparently the issue I am seeing is related?

[1] https://github.com/bazelbuild/continuous-integration/blob/master/buildkite/README.md#faq
[2] https://github.com/bazelbuild/bazel/pull/11436

davido commented:
I guess I don't understand why the test `//src/test/shell/bazel:bazel_coverage_java_jdk14_toolchain_head_test` still failing on the CI, after following the instructions from https://github.com/bazelbuild/bazel/pull/11436#issuecomment-631014072.

@philwo @meteorcloudy any ides?

The test is here:

```python
# Test java coverage with the java_toolchain in the java_tools zip built at head.
[
    sh_test( 
        name = "bazel_coverage_java_jdk" + java_version + "_toolchain_head_test",
        srcs = ["bazel_coverage_java_test.sh"],
        args = [
            # --java_toolchain
            "@local_java_tools//:toolchain",
            # java_tools zip to test
            "src/java_tools_java" + java_version + ".zip",
            # --javabase value
        ] + select({
            "//src/conditions:darwin": ["@openjdk" + java_version + "_darwin_archive//:runtime"],
            "//src/conditions:darwin_x86_64": ["@openjdk" + java_version + "_darwin_archive//:runtime"],
            "//src/conditions:windows": ["@openjdk" + java_version + "_windows_archive//:runtime"],
            "//src/conditions:linux_x86_64": ["@openjdk" + java_version + "_linux_archive//:runtime"],
        }),
        data = [
            ":test-deps",
            "//src:java_tools_java" + java_version + "_zip",
            "//src/test/shell/bazel/testdata:jdk_http_archives_filegroup",
        ],
        tags = [
            "no_windows",
        ],
    )
    for java_version in JAVA_TOOLS_VERSIONS
]

``` 

I disabled it for now: `JAVA_TOOLS_VERSIONS = ("11",)`,
but it would be great to understand why it is failing on the CI.
Is there any possibility to troubleshoot the problem on the CI directly?

davido commented:
`bazel_java14_test` failure on Mac OS X seems to be related:

```
ERROR: /private/var/tmp/_bazel_buildkite/d4b6b1e46f2b3ed9c4c55e002c0dab54/execroot/io_bazel/_tmp/8cb3af0b3725c67c81f5d69657690e6a/root/41c9bd7314f6c8ec46377daf39a1a7ad/external/bazel_tools/tools/jdk/BUILD:66:26: every rule of type java_runtime_alias implicitly depends upon the target '@openjdk12_darwin_archive//:runtime', but this target could not be found because of: no such package '@openjdk12_darwin_archive//': The repository '@openjdk12_darwin_archive' could not be resolved
ERROR: Analysis of target '//java/main:Javac14Example' failed; build aborted: Analysis failed
INFO: Elapsed time: 5.181s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (14 packages loaded, 45 targets configured)
ERROR: Build failed. Not running target
FAILED: Build did NOT complete successfully (14 packages loaded, 45 targets configured)
```

And on Windows:

```
ERROR: C:/b/komsf7d6/execroot/io_bazel/_tmp/a91705879816f2a5a46f5e840d98ba73/root/ogmfis5l/external/bazel_tools/tools/jdk/BUILD:66:26: every rule of type java_runtime_alias implicitly depends upon the target '@openjdk12_windows_archive//:runtime', but this target could not be found because of: no such package '@openjdk12_windows_archive//': The repository '@openjdk12_windows_archive' could not be resolved
ERROR: Analysis of target '//java/main:Javac14Example' failed; build aborted: Analysis failed
INFO: Elapsed time: 4.272s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (15 packages loaded, 58 targets configured)
ERROR: Build failed. Not running target
FAILED: Build did NOT complete successfully (15 packages loaded, 58 targets configured)
```

It's trying to use wrong `openjdk12_os` release:

```
external/bazel_tools/tools/jdk/BUILD:66:26: every rule of type java_runtime_alias implicitly depends upon the target 
'@openjdk12_windows_archive//:runtime', but this target could not be found because of: no such package 
'@openjdk12_windows_archive//': The repository '@openjdk12_windows_archive' could not be resolved
ERROR: Analysis of target '//java/main:Javac14Example' failed; build aborted: Analysis failed
```

davido commented:
On RBE one test is failing with timeout, that doesn't seem to be related to this PR, [link](https://buildkite.com/bazel/bazel-bazel-github-presubmit/builds/6490#3f298dfd-da1e-4b51-b58f-c335ce7a2f52)?

```
(10:34:24) ERROR: /workdir/src/main/java/com/google/devtools/build/lib/BUILD:521:8: Executing genrule //src/main/java/com/google/devtools/build/lib:gen_skylarklibrary failed (Timeout): bash failed: error executing command
```

davido commented:
The `//src/test/shell/bazel:bazel_docgen_test` breakage on RBE seems to be unrelated?

```bash
(12:50:27) INFO: From Executing genrule //:bazel-distfile-tar:
./combine_distfiles_to_tar.sh: 44: [: Linux: unexpected operator
(13:00:04) ERROR: /workdir/src/main/java/com/google/devtools/build/lib/BUILD:521:8: Executing genrule //src/main/java/com/google/devtools/build/lib:gen_skylarklibrary failed (Timeout): bash failed: error executing command

(cd /var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/execroot/io_bazel && \
   exec env - \
     PATH=/bin:/usr/bin:/usr/local/bin \
 /bin/bash -c 'source external/bazel_tools/tools/genrule/genrule-setup.sh; mkdir -p bazel-out/k8-fastbuild/bin/src/main/java/com/google/devtools/build/lib/skylark-lib &&bazel-out/host/bin/src/main/java/com/google/devtools/build/docgen/skydoc_bin bazel-out/k8-fastbuild/bin/src/main/java/com/google/devtools/build/lib/skylark-lib && zip -qj bazel-out/k8-fastbuild/bin/src/main/java/com/google/devtools/build/lib/skylark-library.zip bazel-out/k8-fastbuild/bin/src/main/java/com/google/devtools/build/lib/skylark-lib/*')
  Execution platform: //:rbe_ubuntu1604_java8_platform (failed due to timeout.)
  Generating Starlark documentation...
```

philwo commented:
@davido We've been seeing some test timeouts on RBE, not sure of the cause yet. If you believe it's unrelated, then there's a good chance that it is.

davido commented:
> 
> 
> @davido We've been seeing some test timeouts on RBE, not sure of the cause yet. If you believe it's unrelated, then there's a good chance that it is.

Thanks, @philwo for confirming.

Can someone help to kick of the pipeline for releasing `java_tools` for toolchain JDK14 from this PR?

I'm not sure, how @iirina did it for JDK12, but there is chicken and egg problem. To release `java_tools` for JDK14,
we need this PR. But we also need to add some tests here, that are using `remote_java_tools_javac14`, that we can't
add yet, because such release is not available yet.

https://github.com/bazelbuild/java_tools/releases only has releases for `javac11` with the current version: v9.0.

davido commented:
@lberki @philwo @comius 

Can we move forward with the review here to make sure toolchain for JDK14 is included in upcoming Bazel releases?

Of course, we could wait for general availbility of JDK 15 on 2020/09/15, see: [1], but if we always wait, we would never release anything new here ;-)

[1] https://openjdk.java.net/projects/jdk/15/

comius commented on WORKSPACE:
```diff
@@ -701,6 +701,16 @@ http_archive(
 
+
+# TODO(davido): This is not needed, because jdk compiler module patching is disabled for JDK 14
+http_archive(
```
This is potentially using a different version of Java compiler. Remove it.

davido replied:
Done.

comius commented on src/BUILD:
```diff
@@ -567,7 +567,7 @@ JAVA_TOOLS_DEPLOY_JARS = [
+JAVA_VERSIONS = ("11", "14")
 
 [
     genrule(
```
Split this genrule into Java 11 and Java 14, and remove the (wrong version of) compilers from Java 14.

davido replied:
I am not sure I follow. Are you suggesting not to use the variables, but copy/paste the entire toolchain genrule? But how that would work and scale for the next java toolchains: 15 and 17, ..., 42?

comius replied:
Either no variables or you can add something like JAVA_VERSIONS_NO_JAVAC_JAR = ("14","15","17",...,"42") :)

davido replied:
Done.

comius commented on tools/jdk/BUILD.java_tools:
```diff
@@ -66,6 +66,18 @@ java_toolchain(
+patch_14 = []
+
 # A toolchain that targets java JAVA_LANGUAGE_LEVEL.
 java_toolchain(
```
Here are two dependencies on the old compiler: "javac" parameter and "tools" parameter. In case of Java 14 you can leave the out (the params are optional). 

 They old compiler is probably never used, but the dependency pulls down unnecessary artifacts.

I know I suggested to you before to use variables, however with three different parameters (after the change) it seems nicer to copy the whole target for Java14.

davido commented:
@comius 

> thanks for your amazing effort, and sorry for the delay. It took me some time to thoroughly review your PR. Those should now be the last changes.

I addressed your comments. PTAL.

davido commented:
> Split this genrule into Java 11 and Java 14, and remove the (wrong version of) compilers from Java 14.

I tried that, but the problem is that the default toolchain is still using them:

```python
java_toolchain(
    name = "toolchain",
    bootclasspath = ["@bazel_tools//tools/jdk:platformclasspath"],
    forcibly_disable_header_compilation = 0,
    genclass = [":GenClass"],
    header_compiler = [":Turbine"],
    header_compiler_direct = [":TurbineDirect"],
    ijar = [":ijar"],
    jacocorunner = ":jacoco_coverage_runner_filegroup",
    javabuilder = [":JavaBuilder"],
    javac = [":javac_jar"],
[...]
```

So that it's still needed. I added a comment, that until default toolchain still relies on custom javac compiler it should still be included into `java_tools` for toolchain for Java 14 and even later.

Of course, when we abandon Java 8 and Java 11 (in 5 years?) we could remove custom javac.

davido commented:
@lberki @philwo @comius 

Can we move forward here, and kick off `java_tools` release pipline from this PR, so that `java_tools` for Java 14 is included in the next Bazle release?

For that I filed this issue in `java_tools` project: https://github.com/bazelbuild/java_tools/issues/28.

davido commented:
@jin (or somebody else)

Could you please remove "WIP" label, as this PR is ready now.
